### PR TITLE
fix(codex-ws): fall back to HTTP on pre-first-token websocket drop (#46)

### LINKED ISSUE #46: Fall back gracefully on mid-stream Codex WebSocket failure
## Summary

Handle a Codex WebSocket failure that occurs **mid-stream** (after the first token) by degrading gracefully instead of erroring. Today a WS failure *before* streaming falls back to HTTP transparently, but a failure *mid-stream* surfaces as an error SSE event to the client.

## Why

The pre-stream fallback already proves the HTTP path is a valid backstop. A socket that drops mid-turn (idle eviction race, backend hiccup, network blip) currently turns a recoverable transport failure into a user-visible error, even though the request could be retried over HTTP (or the WS reconnected and continued via `previous_response_id`). This is the one WS failure mode that isn't yet covered.

See `docs/comparison.md` §6 item C.

## Tasks

- [ ] Detect mid-stream WS failure distinctly from a normal terminal event.
- [ ] Decide the safe recovery: (a) reconnect + continue via `previous_response_id` when the turn is resumable, or (b) restart the turn over HTTP. Note that tokens already emitted to the client constrain (b) — spec out the correct behavior (e.g. only fall back if nothing was streamed yet, else surface a clean error).
- [ ] Invalidate the pooled connection on any mid-stream failure (confirm `invalidate_pool_key` is hit on this path).
- [ ] Add tests for mid-stream drop → recovery/clean-error.

## Evidence

- `src/adapters/responses.rs:92-135` (pre-stream fallback; mid-stream surfaces as error SSE at `:554`)
- `src/adapters/codex_ws.rs:110-112` (pool invalidation on error)

## Related

- Comparison: `docs/comparison.md`
- Sibling WS items: live-probe normalization, speculative prewarm

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Closes #46. A Codex WebSocket failure *before the first event reaches the client* now falls back to HTTP transparently, extending the pre-handshake safety net across the send→first-token window. A failure *after* the first event has streamed stays a clean Anthropic `error` SSE (an HTTP restart would duplicate already-streamed tokens).

### Why it was still broken

The pre-stream fallback in `forward()` only caught failures before `forward_websocket` returned. But `Turn::stream` merely *queues* the frame — the connection reader sends it and produces the first event *asynchronously* — so a socket dropping between the send and the first token (idle-eviction race, backend hiccup) surfaced as an `error` SSE on an already-committed 200 stream, never reaching the fallback.

### Changes

- `src/adapters/responses/mod.rs`: `open_ws_turn` now always peeks the first event before committing; new `peek_first_event` + `commit_or_fallback` helpers return `Err` (→ HTTP fallback) on a pre-first-token transport error or empty stream, and buffer/replay the first `Ok` event otherwise. Backend-sent error *events* (429, content-policy) arrive as `Ok` and stream through unchanged. Mid-turn `previous_response_id` resume is a deliberate non-goal.
- Pool invalidation on mid-stream failure (issue task 3) was already handled by `evict()` → `invalidate_pool_key` on every `TurnEnd::Dead` path — confirmed, no change needed.
- `tests/codex_websocket_fallback.rs`: dual WS+HTTP mock upstream on one port; `websocket_drop_before_first_event_falls_back_to_http` and `websocket_drop_after_first_event_surfaces_clean_error`.
- Docs: `docs/m7-codex-websocket.md` (fallback semantics), `docs/comparison.md` §6 item C (→ resolved) + §7 takeaway, `site/src/content/docs/reference/configuration.md` (`websocket` key).

## Milestone / spec

M7 — `docs/m7-codex-websocket.md` (fallback semantics)

## Checklist

- [x] `cargo build` passes
- [x] `cargo test` passes (new behavior is covered; tests run without network/loopback where possible)
- [x] `cargo clippy --all-targets -- -D warnings` clean
- [x] `cargo fmt --all --check` clean
- [x] Source files stay under 500 lines
- [x] English only; matches surrounding style
- [x] Frozen spec in `docs/` updated if this change deviates from it
- [x] User-facing docs updated for behavior/config/endpoint/CLI/provider/model changes — `README.md` / `site/` as applicable (`wiki/` is generated; don't hand-edit)
- [x] Any new GitHub Action is pinned to a full commit SHA

## Notes for reviewers

The key subtlety is the peek-before-commit split: `Turn::stream` queuing is asynchronous relative to the WebSocket send, so the failure window extends past `forward_websocket` returning. Please double check the buffering in `commit_or_fallback` doesn't drop or duplicate the first event on the happy path, and that backend-sent error events (429/content-policy) are correctly distinguished from transport-level failures (both arrive as stream items, but only the latter should trigger fallback).

## Testing

- `cargo fmt --all --check` ✓
- `cargo clippy --all-targets --all-features -- -D warnings` ✓ (0 warnings)
- `RUSTFLAGS=-D warnings cargo test --all-features --workspace` ✓ (all pass, incl. the 3 fallback tests)

---
## Summary by cubic
Fixes a Codex WebSocket edge case: if the socket drops before the first event reaches the client, the turn now falls back to HTTP transparently. Drops after the first event still surface cleanly (SSE error for streaming, 502 for non‑streaming). Fixes #46.

- **Bug Fixes**
  - Always peek the first WS event before committing; `peek_first_event` + `commit_or_fallback` fall back on a pre-first-event transport error or empty stream, and buffer/replay the first `Ok` event otherwise.
  - Backend error events still stream through; only transport failures trigger fallback. Mid-stream drops: streaming → Anthropic `error` SSE; non‑streaming → 502 gateway error (no HTTP retry).
  - Tests: dual WS+HTTP upstream harness; added empty‑stream unit test for `commit_or_fallback` and a non‑streaming mid‑stream drop test; tightened assertion to exactly one HTTP POST on fallback; avoid busy‑spin in `request_is_websocket` by backing off on partial `peek`; make mock `ws.send` failures loud.
  - Docs: correct “first token” → “first event” across comments and docs; document non‑streaming JSON error asymmetry; resolve comparison §7 contradiction while marking item C resolved by #46 and keeping #45 open; config docs note “falls back on any failure before the first event.”

<sup>Written for commit c2b040c004daa2c0b6827d630d851e56ca519688. Summary will update on new commits.</sup>

### CODE REVIEW

File: tests/codex_websocket_fallback.rs [hunk 1]
```diff
@@ -181,3 +195,231 @@ async fn websocket_handshake_failure_falls_back_to_http() {
...
+async fn request_is_websocket(socket: &TcpStream) -> bool {
+    let mut head = [0u8; 4];
+    loop {
+        match socket.peek(&mut head).await {
+            Ok(0) | Err(_) => return false,
+            Ok(n) if n >= 4 => return &head == b"GET ",
+            Ok(_) => continue, // need more bytes before deciding
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply tests/codex_websocket_fallback.rs:269]: @gemini-code-assist 좋은 지적 감사합니다. `request_is_websocket`에서 부분 읽기(4바이트 미만) 시 `peek`이 버퍼된 바이트를 그대로 즉시 반환해 발생하는 busy loop을 해당 arm에서 10ms 백오프(`tokio::time::sleep`)로 수정했습니다 (3ac2294).

File: src/adapters/responses/mod.rs [hunk 2]
```diff
@@ -554,9 +600,10 @@ fn stream_events_response(
 }
 
 /// Collect the full websocket event stream into a single Anthropic message for a
-/// non-streaming client. A mid-stream transport error returns a gateway error
-/// instead of presenting partial output as a successful response. `buffered` is
-/// the peeked first event, if any.
+/// non-streaming client. The first event was peeked in [`open_ws_turn`] (a
+/// pre-first-event failure already fell back to HTTP), so a transport error here
+/// is mid-stream: return a gateway error instead of presenting partial output as
+/// a successful response. `buffered` is the replayed first event, if any.
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/adapters/responses/mod.rs:606]: @gemini-code-assist 지적 감사합니다. `json_events_response` 주석에 스트리밍/논스트리밍 에러 처리 비대칭을 명시했습니다. 백엔드가 보낸 에러 이벤트가 200으로 조용히 반환되는 silent-failure는 pre-existing 동작(HTTP json 경로와 공유)이라 이 PR 범위 밖으로 두고 별도 이슈 #113으로 추적합니다 (b74a2a8).
